File: ganilib-core/src/main/java/com/gani/lib/dialog/GRestDialogProgress.java
//package com.gani.lib.dialog;
//
//import android.content.Intent;
//import android.os.Bundle;
//
//import com.gani.lib.http.GImmutableParams;
//import com.gani.lib.http.GRestCallback;
//import com.gani.lib.http.GRestResponse;
//import com.gani.lib.http.HttpAsync;
//import com.gani.lib.http.HttpMethod;
//
//import org.map.JSONException;
//
//import java.io.Serializable;
//
//public class GRestDialogProgress extends GDialogProgress {
//  private static final String ARG_METHOD = "method";
//  private static final String ARG_URL = "url";
//  private static final String ARG_PARAMS = "params";
//  private static final String ARG_CALLBACK = "callback";
//
//  private HttpAsync httpRequest;
//
//  public static Intent intent(HttpMethod method, String url, GImmutableParams params, Callback callback) {
//    return Companion.intentBuilder(GRestDialogProgress.class)
//        .withArg(ARG_METHOD, method)
//        .withArg(ARG_URL, url)
//        .withArg(ARG_PARAMS, params)
//        .withArg(ARG_CALLBACK, callback)
//        .getIntent();
//  }
//
//  @Override
//  protected void onCreate(Bundle savedInstanceState) {
//    super.onCreate(savedInstanceState);
//
//    HttpMethod method = (HttpMethod) args().getSerializable(ARG_METHOD);
//    String url = args().getString(ARG_URL);
//    GImmutableParams params = args().getParams(ARG_PARAMS);
//    final Callback callback = (Callback) args().getSerializable(ARG_CALLBACK);
//
//    httpRequest = method.async(url, params, new GRestCallback(this) {
////      @Override
////      protected void onJsonSuccess(GRestResponse r) throws JSONException {
////        callback.onJsonSuccess(GRestDialogProgress.this, r);
////      }
//
//      @Override
//      protected void onRestResponse(GRestResponse r) throws JSONException {
//        callback.onRestResponse(GRestDialogProgress.this, r);
//      }
//    }).execute();
//  }
//
//  @Override
//  protected void onCancel() {
//    httpRequest.cancel();
//  }
//
//
//
//  public static abstract class Callback implements Serializable {
//    // To be overridden
//    protected void onRestResponse(GRestDialogProgress dialogProgress, GRestResponse r) throws JSONException {
//      GRestCallback.displayMessage(r);
//    }
//  }
//}

File: ganilib-core/src/main/java/com/gani/lib/dialog/GRestDialogProgress.kt
package com.gani.lib.dialog

import android.content.Intent
import com.gani.lib.http.*
import com.gani.lib.screen.GActivity
import com.gani.lib.screen.GFragment
import com.gani.lib.screen.GScreenContainer
import java.io.Serializable

class GRestDialogProgress : GDialogProgress() {
    override fun createNewIntentFragment(): GFragment {
        return ContentFragment()
    }



    companion object {
        private val ARG_METHOD = "method"
        private val ARG_URL = "url"
        private val ARG_PARAMS = "params"
        private val ARG_CALLBACK = "callback"

        fun intentForUrl(url: String, method: Rest, params: GParams.Default?, callback: Callback): Intent {
            return intentBuilder(GRestDialogProgress::class)
                    .withArg(ARG_METHOD, method)
                    .withArg(ARG_URL, url)
                    .withArg(ARG_PARAMS, params)
                    .withArg(ARG_CALLBACK, callback)
                    .intent
        }

        fun intentForPath(path: String, method: Rest, params: GParams.Default?, callback: Callback): Intent {
            return intentForUrl("${GHttp.instance.host}${path}", method, params, callback)
        }
    }



    interface Callback : Serializable {
        fun onRestResponse(fragment: GFragment, response: GRestResponse)
    }



    class ContentFragment : GFragment() {

        private var httpRequest: HttpAsync? = null

        override fun initContent(activity: GActivity, container: GScreenContainer) {
            val method = args[ARG_METHOD] as Rest
            val url = args[ARG_URL] as String
            val params = args[ARG_PARAMS] as? GParams.Default
            val callback = args[ARG_CALLBACK] as Callback

            httpRequest = method.asyncUrl(url, params, object : GRestCallback.Default(this@ContentFragment) {
                override fun onRestResponse(response: GRestResponse) {
                    super.onRestResponse(response)

                    callback.onRestResponse(this@ContentFragment, response)
                    gActivity?.finish()
                }
            }).execute()
        }

        override fun onDestroy() {
            super.onDestroy()
            httpRequest?.cancel()
        }

    }
}
